system/src/ble_listening_mode_handler.cpp

```diff
@@ -34,112 +34,103 @@ using namespace particle::system;
-    memcpy(&ctrlReqAdvData_[ctrlReqAdvDataLen_], devName, strlen(devName));
-    ctrlReqAdvDataLen_ += strlen(devName);
+    CHECK_TRUE(tempAdvData.append(strlen(devName) + 1), SYSTEM_ERROR_NO_MEMORY);
+    CHECK_TRUE(tempAdvData.append(BLE_SIG_AD_TYPE_COMPLETE_LOCAL_NAME), SYSTEM_ERROR_NO_MEMORY);
+    CHECK_TRUE(tempAdvData.append((uint8_t*)devName, strlen(devName)), SYSTEM_ERROR_NO_MEMORY);
 
     uint16_t platformID = PLATFORM_ID;
     uint16_t companyID = PARTICLE_COMPANY_ID;
 
     // Manufacturing specific data
-    ctrlReqAdvData_[ctrlReqAdvDataLen_++] = sizeof(platformID) + sizeof(companyID) + 1;
-    ctrlReqAdvData_[ctrlReqAdvDataLen_++] = BLE_SIG_AD_TYPE_MANUFACTURER_SPECIFIC_DATA;
-    memcpy(&ctrlReqAdvData_[ctrlReqAdvDataLen_], (uint8_t*)&companyID, 2);
-    ctrlReqAdvDataLen_ += sizeof(companyID);
-    memcpy(&ctrlReqAdvData_[ctrlReqAdvDataLen_], (uint8_t*)&platformID, sizeof(platformID));
-    ctrlReqAdvDataLen_ += sizeof(platformID);
+    CHECK_TRUE(tempAdvData.append(sizeof(platformID) + sizeof(companyID) + 1), SYSTEM_ERROR_NO_MEMORY);
+    CHECK_TRUE(tempAdvData.append(BLE_SIG_AD_TYPE_MANUFACTURER_SPECIFIC_DATA), SYSTEM_ERROR_NO_MEMORY);
+    CHECK_TRUE(tempAdvData.append((uint8_t*)&companyID, 2), SYSTEM_ERROR_NO_MEMORY);
+    CHECK_TRUE(tempAdvData.append((uint8_t*)&platformID, sizeof(platformID)), SYSTEM_ERROR_NO_MEMORY);
 
     // Particle Control Request Service 128-bits UUID
-    ctrlReqSrData_[ctrlReqSrDataLen_++] = sizeof(BLE_CTRL_REQ_SVC_UUID) + 1;
-    ctrlReqSrData_[ctrlReqSrDataLen_++] = BLE_SIG_AD_TYPE_128BIT_SERVICE_UUID_COMPLETE;
-    memcpy(&ctrlReqSrData_[ctrlReqSrDataLen_], BLE_CTRL_REQ_SVC_UUID, sizeof(BLE_CTRL_REQ_SVC_UUID));
-    ctrlReqSrDataLen_ += sizeof(BLE_CTRL_REQ_SVC_UUID);
+    CHECK_TRUE(tempSrData.append(sizeof(BLE_CTRL_REQ_SVC_UUID) + 1), SYSTEM_ERROR_NO_MEMORY);
+    CHECK_TRUE(tempSrData.append(BLE_SIG_AD_TYPE_128BIT_SERVICE_UUID_COMPLETE), SYSTEM_ERROR_NO_MEMORY);
+    CHECK_TRUE(tempSrData.append(BLE_CTRL_REQ_SVC_UUID, sizeof(BLE_CTRL_REQ_SVC_UUID)), SYSTEM_ERROR_NO_MEMORY);
+
+    ctrlReqAdvData_ = std::move(tempAdvData);
+    ctrlReqSrData_ = std::move(tempSrData);
     return SYSTEM_ERROR_NONE;
 }
 
 int BleListeningModeHandler::cacheUserConfigurations() {
     LOG_DEBUG(TRACE, "Cache user's BLE configurations.");
+    CHECK_FALSE(exited_, SYSTEM_ERROR_INVALID_STATE);
+
+    Vector<uint8_t> tempAdvData(BLE_MAX_ADV_DATA_LEN);
+    Vector<uint8_t> tempSrData(BLE_MAX_ADV_DATA_LEN);
 
     // Advertising data set by user application
-    preAdvDataLen_ = hal_ble_gap_get_advertising_data(nullptr, 0, nullptr);
-    CHECK(preAdvDataLen_);
-    if (preAdvDataLen_ > 0) {
-        preAdvData_.reset(new(std::nothrow) uint8_t[preAdvDataLen_]);
-        CHECK_TRUE(preAdvData_, SYSTEM_ERROR_NO_MEMORY);
-        hal_ble_gap_get_advertising_data(preAdvData_.get(), preAdvDataLen_, nullptr);
-    }
+    size_t len = hal_ble_gap_get_advertising_data(tempAdvData.data(), BLE_MAX_ADV_DATA_LEN, nullptr);
```

Comment:
`ssize_t` or even better:

```cpp
size_t len = CHECK(hal_ble_gap_get_advertising_data(tempAdvData.data(), BLE_MAX_ADV_DATA_LEN, nullptr));
```

Reply:
Same in https://github.com/particle-iot/device-os/pull/1882/commits/fc3555e8e938d92fd5f954ae3665d11d5e968e25#diff-79e0c2ab765d993495c5c34f5f338f74R101